staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -402,16 +466,50 @@ func (s *store) List(ctx context.Context, key, resourceVersion string, pred stor
 	if !strings.HasSuffix(key, "/") {
 		key += "/"
 	}
-	getResp, err := s.client.KV.Get(ctx, key, clientv3.WithPrefix())
+	keyPrefix := key
+
+	// set the appropriate clientv3 options to filter the returned data set
+	options := make([]clientv3.OpOption, 0, 4)
+	if pred.Limit > 0 {
+		options = append(options, clientv3.WithLimit(pred.Limit))
+	}
+	var returnedRV int64
+	switch {
+	case len(pred.From) > 0:
+		fromKey, fromRV, err := decodeFrom(pred.From, keyPrefix)
+		if err != nil {
+			return err
+		}
+
+		options = append(options, clientv3.WithRange(key+"\xFF"))
```

Comment:
decodeFrom performs that validation already.  The from key is a subset of the range, so there is the implicit behavior of the internal structure of the tree.  We already handle path traversal, and the mechanism here depends on this being a sorted range.  I think at some point we would offer "startFrom" (I think `From` in this PR would be renamed to `Continue` and be opaque, but `From` would be a generic tool that would have to follow a recognizable pattern).